### ISSUE TITLE
Advanced searches including Boolean NOT returning too few hits

### ISSUE THREAD

**Issue by andrew-morrison** [archivesspace/archivesspace#1699]
## Expected Behavior
Via the public user interface, entering a query such as `golf NOT tennis` all in the first search box should produce the same results as entering `golf` in the first row, adding a second row, selecting "Not" and entering `tennis` in the second row.

## Current Behavior
Instead, the number of hits is always lower when such a NOT query built as an advanced search, using multiple rows, than if entered in one row. At time of testing, the following results were obtained on a few different ArchivesSpace PUIs:
```
                     Bodleian    Harvard     Yale    Sandbox
golf                 8           327         286     5
tennis               27          378         348     0
golf AND tennis      1           23          9       0
golf OR tennis       34          682         625     5
golf NOT tennis      7           304         277     5
advanced search AND  1           23          9       0
advanced search OR   34          682         625     5
advanced search NOT  0           292         7       0
```
The last line is the problem. The number of hits should be the number returned by `golf` minus the number returned by `golf AND tennis`. That is the case for a simple `golf NOT tennis` search, but is not when the same query is built using multiple rows.

In that scenario, the public user interface `set_up_advanced_search()` method builds a JSON data structure, conforming to the `advanced_query` model, as below:
```json
{
  "query": {
    "op": "AND",
    "subqueries": [
      {
        "field": "publish",
        "value": "true",
        "negated": false,
        "literal": false,
        "jsonmodel_type": "field_query"
      },
      {
        "op": "AND",
        "subqueries": [
          {
            "field": "types",
            "value": "pui",
            "negated": false,
            "literal": false,
            "jsonmodel_type": "field_query"
          },
          {
            "op": "AND",
            "subqueries": [
              {
                "op": "AND",
                "subqueries": [
                  {
                    "field": "keyword",
                    "value": "tennis",
                    "negated": true,
                    "literal": false,
                    "jsonmodel_type": "field_query"
                  }
                ],
                "jsonmodel_type": "boolean_query"
              },
              {
                "op": "AND",
                "subqueries": [
                  {
                    "op": "AND",
                    "subqueries": [
                      {
                        "field": "keyword",
                        "value": "golf",
                        "negated": false,
                        "literal": false,
                        "jsonmodel_type": "field_query"
                      }
                    ],
                    "jsonmodel_type": "boolean_query"
                  }
                ],
                "jsonmodel_type": "boolean_query"
              }
            ],
            "jsonmodel_type": "boolean_query"
          }
        ],
        "jsonmodel_type": "boolean_query"
      }
    ],
    "jsonmodel_type": "boolean_query"
  },
  "jsonmodel_type": "advanced_query"
}
```
That is sent off to the backend, which converts it into Solr query syntax as follows:
```
(publish:(true) AND (types:(pui) AND ((-fullrecord:(tennis) AND *:*) AND ((fullrecord:(golf))))))
```
The `AND *:*` bit has been added because, according to a comment in the code, _"Solr doesn't allow purely negative expression groups, so we add a match all query to compensate when we hit one of these."_ But that is not the case with the extended DisMax query parser being used. Even as far back as the [4.10 version of the Solr documentation](http://archive.apache.org/dist/lucene/solr/ref-guide/apache-solr-ref-guide-4.10.pdf), it says eDisMax "supports pure negative nested queries: queries such as +foo (-foo) will match all documents."

Running the above through the Solr debugQuery mode reveals that it parses and converts it into:
```
+(+publish:true +(+types:pui +(+(-fullrecord:tennis +DisjunctionMaxQuery((fullrecord:*\\:* | (four_part_id:*\\:*)^50.0 | (title:*\\:*)^25.0))) +fullrecord:golf)))
```
The four_part_id and title fields are added because they're specified in the qf (Query Fields) parameter in solrconfig.xml.

## Possible Solution
If the query that the backend sends to the Solr is modified to:
```
(publish:(true) AND (types:(pui) AND ((-fullrecord:(tennis)) AND ((fullrecord:(golf))))))
```
Then Solr parses that as:
```
+(+publish:true +(+types:pui +(+(-fullrecord:tennis +MatchAllDocsQuery(*:*)) +fullrecord:golf)))
```
Now the results returned are as expected. `MatchAllDocsQuery(*:*)` is the eDisMax query parser doing what the ArchivesSpace backend is attempting: converting the negative subquery into something that will produce a set which can be combined with sets found by the other subqueries. But it is better-placed to do so.

This can be achieved by commenting out these lines in solr.rb:

https://github.com/archivesspace/archivesspace/blob/0fd76a2ccacea10451575aecb56fcd3ce9bc8dcd/backend/app/model/solr.rb#L50-L52

However, that was added in https://github.com/archivesspace/archivesspace/commit/d4473b71a4ebe8c0f7030ea3869e435584965cb5 by @marktriggs to achieve "Correct handling of pure NOT queries". I cannot find an issue linked to it, so do not know precisely what scenario it was developed to improve, but the spec added in the same commit tests for Solr query which also results in fewer hits than simple maths suggests is correct.

Note that this problem does not occur when a _something NOT something else_ query is built in the staff interface's advanced search. That is because it generates a much simpler advanced_query JSON object:
```
{
  "query": {
    "op": "AND",
    "subqueries": [
      {
        "field": "keyword",
        "value": "golf",
        "comparator": "contains",
        "jsonmodel_type": "field_query",
        "negated": false,
        "literal": false
      },
      {
        "field": "keyword",
        "value": "tennis",
        "comparator": "contains",
        "negated": true,
        "jsonmodel_type": "field_query",
        "literal": false
      }
    ],
    "jsonmodel_type": "boolean_query"
  },
  "jsonmodel_type": "advanced_query"
}
```
Not just is there no need to filter out the unpublished and non-PUI records, but the two field_query are siblings, and only one is negated, so `if advanced_query['subqueries'].all? {|subquery| subquery['negated']}` is not true. However, if you select Not in the first drop-down in the staff interface search form, it returns too few hits, so I still think it is the backend code that needs to be changed.

## Steps to Reproduce (for bugs)
1. Paste the following on to the end of the URL of your PUI:
* /search?utf8=✓&op[]=&q[]=golf+NOT+tennis&field[]=&from_year[]=&to_year[]=
* /search?q[]=golf&op[]=&field[]=keyword&from_year[]=&to_year[]=&q[]=tennis&op[]=NOT&field[]=keyword&from_year[]=&to_year[]=
2. Compare the number of hits returned for each (choose different keywords if neither return anything, but they should be a pair that is likely to overlap in at least one record.)

## Context
Looking at our access logs, only one person has used Boolean NOT in this way in the fortnight since launching our ArchivesSpace system to the public. But it should nevertheless return correct results.

Possibly this is what was being reported in https://archivesspace.atlassian.net/browse/ANW-753.

## Your Environment
Tested in 2.5.1 and 2.7.0, with internal Solr 4.10 and external Solr 7.7. Probably affects all ArchivesSpace v2 systems.

**Comment by andrew-morrison** [comment]
Note that since raising this issue I have fixed the Bodleian system using a plug-in, so it now returns the correct number of hits (currently 8) whether using multiple rows or all in one:

![simple_search_not](https://user-images.githubusercontent.com/33721187/73679518-74726b00-46b2-11ea-936c-ae8c75db1bca.png)

![advanced_search_not](https://user-images.githubusercontent.com/33721187/73679521-76d4c500-46b2-11ea-94f7-5a203f73fc0c.png)

I'm submitting the same fix as a pull request

**Comment by andrew-morrison** [comment]
@lorawoodford: I believe this issue can be closed, as it has been fixed by #1791, and the side-effect on typeahead searches has been fixed by #1881 . I've tested again today on test.archivesspace.org.

**Comment by lorawoodford** [comment]
Thanks @andrew-morrison
